v3 update

christopher-johnson opened the pull request:
this includes:
- refactor and restructure of producer driver
- replace POJO models with builder models
- removal of global configuration and addition of default parameter values
- addition of format parameter that triggers either v2 or v3 serialization format (v3 is default)
- complete v3 converter implementation that includes multilingual metadata labels at manifest and structure levels
- initial testing and validation of v2 and v3 monograph, multi-volume, manuscript and HSP types in Mirador 3
- update of Gradle and build tooling to latest

Note: structure level metadata is not rendered in Mirador 3 (yet).  Please advise on HSP and structure metadata implementation details.

**v3 Top Metadata View**
![Screen Shot 2022-02-09 at 7 44 23 AM (3)](https://user-images.githubusercontent.com/5990301/153218763-f8675d0f-82e1-4d36-81bf-4517fc5115f8.png)

**v3 Structures View**
![Screen Shot 2022-02-09 at 7 44 06 AM (3)](https://user-images.githubusercontent.com/5990301/153218965-98940b76-dfdc-46bd-b33d-033f5ce8b8f8.png)

Commit:
rewrite for updates

Commit:
refactor global config from producer
fix tests

Commit:
update models with builders
add v3 models (wip)

Commit:
add defaults to builders
remove old models

Commit:
implement v3 converter (WIP)

Commit:
metadata implementation (WIP)

Commit:
metadata implementation (WIP)

Commit:
structure metadata (WIP)

Commit:
multilanguage labels

Commit:
adds abstract producer tests
set multivalue metadata

Commit:
structures final

Commit:
update README

Commit:
remove commented exclusions

Commit:
remove duplicate license headers

Commit:
remove unnecessary setters and getters

Commit:
remove test output files

lutzhelm commented on converter/src/main/java/de/ubleipzig/iiifproducer/converter/ConverterVersion3.java:
```diff
@@ -0,0 +1,262 @@
+                .type(MANIFEST)
+                .viewingDirection("left-to-right")
+                .behavior(behaviors)
+                .rights(domainLicense)
```
`rights` cannot simply be a single default license. Public Domain Mark 1.0 is only applicable for manuscripts and historical prints when there are no other copyright restrictions. But there are several digitization projects at UBL that produce digital objects that are still in copyright. They often have multiple licenses linked in v2. The content of `rights` must reflect that.

To further complicate things, [`rights`](https://iiif.io/api/presentation/3.0/#rights) must be a single string value with either a creative commons or a rightsstatements.org URI; an array like v2 `license` is not allowed.

My suggestion would be to
* filter v2 `licenses` for the first entry that is either a creative commons or a rightsstatements.org license and use it for `rights`
* add _every_ license found in v2 `license` including the one that is used in `rights` as a HTML link at the end of the content of `requiredStatement`, e.g. `https://www.ub.uni-leipzig.de/recherche/katalog-information/nutzungshinweis-rechte-vorbehalten-freier-zugang/` becomes `<a href="https://www.ub.uni-leipzig.de/recherche/katalog-information/nutzungshinweis-rechte-vorbehalten-freier-zugang/">https://www.ub.uni-leipzig.de/recherche/katalog-information/nutzungshinweis-rechte-vorbehalten-freier-zugang/</a>`, either wrapped within a single `<div/>` or with necessary `<br/>`s for line breaks between all URIs and the main content of `requiredStatement`.
  * an alternative would be to reproduce those URIs whithin the manifest metadata but this would result in them not showing up in the rights panel in Mirador 3, for example.

Before you implement any of there suggestions: we are still waiting for feedback from Björn on what he thinks is best with respect to UBL manifests.

christopher-johnson replied:
implemented new required statement as suggested.  note that if license is included in attribution, then Mirador will show the value in both the "attribution" and "license" sections of the rights interface
![Screen Shot 2022-03-09 at 9 01 34 AM (3)](https://user-images.githubusercontent.com/5990301/157457040-837fbfa8-e01e-40c1-9f77-dbce333fe4ab.png)

lutzhelm commented on converter/src/main/java/de/ubleipzig/iiifproducer/converter/ConverterVersion3.java:
```diff
@@ -0,0 +1,262 @@
+        final List<String> behaviors = new ArrayList<>();
+        behaviors.add("paged");
+
+        final Map<String, List<String>> label = buildLabelMap("Attribution", ENGLISH);
+        final Map<String, List<String>> value = buildLabelMap(domainAttribution, ENGLISH);
+        final MetadataVersion3 requiredStatement = MetadataVersion3.builder()
+                .label(label)
+                .value(value)
+                .build();
```
Required statement can't be a single default statement. It should reproduce the content of v2 `attribution`. There are several projects at UBL that require different statements because they are still in copyright.

christopher-johnson replied:
see above new requiredStatement implementation

lutzhelm commented on converter/src/main/java/de/ubleipzig/iiifproducer/converter/StructureBuilderVersion3.java:
```diff
@@ -0,0 +1,304 @@
+        //finally merge all substructures into top structure
+        final Optional<Item> topStructure = newStructures.stream().filter(
+                s -> s.getId().contains("LOG_0000")).findAny();
+        if (topStructure.isPresent()) {
+            final Map<String, List<String>> topStructurelabelMap = buildLabelMap("Contents", ENGLISH);
+            topStructure.get().setLabel(topStructurelabelMap);
```
Converting the v2 top level structure (which is usually not displayed by clients) into an own structure creates an additional navigation entry in both Universal Viewer and Mirador 3. Both viewers could display manifests without such a top level Range correctly, but creating a top level structure is in what's suggested in the [v3 cookbook recipe](https://iiif.io/api/cookbook/recipe/0024-book-4-toc/), so I'm not sure how to proceed.

For now, it might be sufficient to add an additional label/value pair with for German and "Inhalt".

![Mirador 3](https://user-images.githubusercontent.com/20013890/156564400-0f54d7e8-e6d3-46d7-aa65-ce1a8ee498d9.png)
![Universal Viewer](https://user-images.githubusercontent.com/20013890/156564405-59d0e156-75f1-4f4d-9fcf-84ba1a9a1f1a.png)

christopher-johnson replied:
implemented "de" label ("Inhalt") for top structure as suggested
![Screen Shot 2022-03-09 at 9 05 34 AM (3)](https://user-images.githubusercontent.com/5990301/157457452-c094a102-c385-4c72-8d87-ec1d8b89f332.png)

lutzhelm commented on model/src/main/java/de/ubleipzig/iiifproducer/model/v3/BodyVersion3.java:
```diff
@@ -0,0 +1,51 @@
+@Getter
+@AllArgsConstructor
+@JsonPropertyOrder({"id", "type", "format", "service", "height", "width"})
+public class BodyVersion3 {
+    @JsonProperty("format")
+    private String format;
+    @JsonProperty
+    private Integer height;
+    @JsonProperty("id")
+    private String id;
+    @JsonProperty
+    private Map<String, List<String>> label;
+    @JsonProperty
+    private List<ServiceVersion3> service;
+    @JsonProperty("type")
+    private String type;
+    @JsonProperty
+    private Integer width;
+}
```
For the HSP manuscript cataloges, `Canvas` requires a `seeAlso` attribute containing an ALTO XML entry. It should be transformed from the v2 manifest. If no value is present, it should be left out for serialization.

lutzhelm commented on model/src/main/java/de/ubleipzig/iiifproducer/model/v3/BodyVersion3.java:
```diff
@@ -0,0 +1,51 @@
+@Builder
+@Setter
+@Getter
+@AllArgsConstructor
```
If `seeAlso` is empty; see comment below.
```suggestion
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
```

christopher-johnson replied:
all empty and null values are now excluded by the mapper

lutzhelm reviewed (changes requested):
Thank you very much, @christopher-johnson! Overall, the changes look very promising, although we could not review every single change line by line.

There are some issues regarding the version 3 manifests, primerily regarding the `rights` and `requiredStatement` properties.

> Note: structure level metadata is not rendered in Mirador 3 (yet). Please advise on HSP and structure metadata implementation details.

Noted. We became aware of that as soon as we tested those manifests with M3. Thank you.

Commit:
implements homepage and requiredStatement
set serialization options on mapper
adds de labels for Content and Manifest Type
refactor utility functions
adds seeAlso to canvases

Commit:
adds converterUtils tests
strip html from and new lines from attribution to avoid multiple line breaks

christopher-johnson commented on converter/src/main/java/de/ubleipzig/iiifproducer/converter/ConverterVersion3.java:
```diff
@@ -0,0 +1,262 @@
+                .build();
+        seeAlso.add(viewerReference);
+        List<String> filteredRelated = related.stream()
+                .filter(r -> !katalogId.equals(r) && !viewerId.equals(r)).collect(Collectors.toList());
```
implemented manifest type of "HSP" to provide conditionality for homepages. added homepage when not HSP type e.g.
```   
"homepage" : [
    {
      "id" : "https://katalog.ub.uni-leipzig.de/urn/urn:nbn:de:bsz:15-0008-232267",
      "type" : "Text",
      "format" : "text/html"
    },
    {
      "id" : "https://digital.ub.uni-leipzig.de/object/viewid/0000004595",
      "type" : "Text",
      "format" : "text/html"
    }
  ], 
```

Commit:
no logo if HSP

christopher-johnson commented on converter/src/main/java/de/ubleipzig/iiifproducer/converter/ConverterVersion3.java:
```diff
@@ -0,0 +1,237 @@
+        final String rights = manifest.getLicense().stream().findFirst().orElse(null);
+
+        //build Logo
+        final ManifestVersion3.Logo logo = ManifestVersion3.Logo.builder()
+                .id(domainLogo)
+                .type(IMAGE)
+                .build();
```
ok.  added a check to only add logo if not an HSP manifest type
